salt/pillar/__init__.py

```diff
@@ -623,6 +616,29 @@ def __gather_avail(self):
             avail[saltenv] = self.client.list_states(saltenv)
         return avail
 
+    def __gen_functions(self, context=None, functions=None, force=False):
+        """
+        Generates the minion_mods loader object using the correct opts, context
+        and pillar values in the correct scenarios
+        """
+        if not force and functions is not None:
+            return functions
+
+        # This is some historical cruft that I don't want to pretend to try to
+        # understand. Keeping compatible for the sake of compatibility.
+        # Origin: 340aed8049b885c0736681e925cf7210ede0f961
+        opts = self.opts
+        if self.raw_opts.get("file_client", "") == "local":
+            opts = self.raw_opts
+            # Ensure we keep the partially-rendered pillar regardless of which
+            # opts we're using. Copy pillar into the alternative opts.
+            if "pillar" in self.opts:
+                opts["pillar"] = self.opts["pillar"]
```

Comment:
What if there is already "pillar" in opts, wouldn't self.opts["pillar"] be writing that over?

Reply:
Yes, probably. The whole "which opts should we use" thing here is an utter mess. It's very unclear what's going on regarding mutability and correctness of where pillar should be preserved and not. I can't really say for sure what the semantics of this should be, but if we remove this statement the thing that this PR attempts to fix will not work. The only reason this branch even exists is to not break compatibility with this ancient nonsense that makes no sense to me: https://github.com/saltstack/salt/commit/340aed8049b885c0736681e925cf7210ede0f961#diff-8b80a647365268f00563c1c4f3e787812a7ad3e07744c7fd71ffc21c7c75b4afR105
I could just drop this entirely, but seeing as there are no tests I have no idea if we'll break anything.

Another option here is to `self.raw_opts.copy()` but that then changes the semantics of the `opts` passed into the `LazyLoader` that we create on line 640.
The way I see it: the whole `opts` thing with respect to mutability and ownership is a joke that is beyond fixing, at least until someone states for definite how it should be handled.

I'm happy to change this to however will be the most compatible/sensible, but ultimately this is extremely fragile code with so many untested edge cases that people probably rely on, that something will inevitably break.

Reply:
Let me reach out to some other core team members and see what they think.

Reply:
@s0undt3ch or @dwoz any opinions here? ^

Reply:
I would like to see pillar data removed from opts completely. I know @s0undt3ch has made some attempts to make opts immutable in the past.

Reply:
> I would like to see pillar data removed from opts completely. I know @s0undt3ch has made some attempts to make opts immutable in the past.

Yes, it's immutable, unless we copy it.
I'm 1000% with keeping opts for what they are supposed to provide, opts, just opts, and remove all runtime additions to it, in this case, pillar.

If the user provides pillar data in opts, then sure, those live in opts, but those should be merged against all pillar data, and not the other way around.

Reply:
@frebib I've created a patch with a different path to solve this issue. Can you please test my branch and see if it fixes the problem you are seeing? #64043

Reply:
Apologies, I'm not sure how but I missed your comment here :(
I'll test this thoroughly tommorrow, but at least your PR passes my test with some minor tweaks (commented on the PR)
```diff
diff --git tests/pytests/unit/pillar/test_pillar.py tests/pytests/unit/pillar/test_pillar.py
index d375a37d47..75bd554a34 100644
--- tests/pytests/unit/pillar/test_pillar.py
+++ tests/pytests/unit/pillar/test_pillar.py
@@ -1,4 +1,7 @@
+from unittest.mock import patch
+
 import pytest
+
 import salt.loader
 import salt.pillar
 from salt.utils.odict import OrderedDict
@@ -122,3 +125,30 @@ def test_pillar_envs_path_substitution(env, temp_salt_minion, tmp_path):

     # The __env__ string in the path has been substituted for the actual env
     assert pillar.opts["pillar_roots"] == expected
+
+
+def test_ext_pillar_dunder_in_modules_in_pillar(temp_salt_minion):
+    """
+    Test that ext_pillar is available in __pillar__ inside execution modules
+    during pillar render when using ext_pillar_first=true
+    """
+    opts = temp_salt_minion.config.copy()
+    opts["ext_pillar_first"] = True
+
+    grains = salt.loader.grains(opts)
+    pillar = salt.pillar.Pillar(opts, grains, temp_salt_minion.id, "base")
+    # Pillar should be empty to start with
+    assert pillar.functions.pack["__pillar__"] == {}
+
+    ext_value = {"ext": "some ext value"}
+    pil_value = {"pillar": "some pillar value"}
+    combined = dict(**ext_value, **pil_value)
+    with (
+        patch.object(pillar, "ext_pillar", return_value=(ext_value, [])),
+        patch.object(pillar, "render_pillar", return_value=(pil_value, [])),
+    ):
+        compiled = pillar.compile_pillar()
+        assert compiled == combined
+
+    # Loader should pack the opts pillar dict from the ext_pillar() call
+    assert pillar.functions.pack["__pillar__"] == combined
```

I'm hopeful that your fix is an improvement over mine. Aside from the obvious hackery I had to go through to make my PR even work, the biggest problem is that somehow mine incurs a ~7 second slowdown reloading all of the modules. I'll measure how your change performs tomorrow and whether it works, but I'm hopeful, although with the ridiculous size of our pillar, I do worry that all of the `dict.update()` calls will be quite slow. We'll see